### PR TITLE
add claims and regex policy selector

### PR DESCRIPTION
**Pull request by butonic**
Using the proxy config file, it is now possible to let let the IdP determine the routing policy by sending an `ocis.routing.policy` claim or by matching username, email or id using regexes. 

- [x] tests
- [x] update changelog
- [ ] update docs

This PR can stand on its own. But the proxy and the docs need an overhaul in subsequent PRs:
- [ ] Configuring the policy selector is only possible via the config file
- [ ] The proxy should always use the cs3 api. get rid of configuration choices here
- [ ] We should use the cs3 api to authenticate requests ... yes it would mean more latency when authenticating, but we can cache the token afterwards. It would remove the duplicate code in the proxy and in the auth providers. This should allow us to actually get rid of several middlewares in here.
- [ ] when moving the oidc auth into reva we can do the claims mapping there, including custom porperties like `ocis.routing.policy` or `ocis.id`

### CODE REVIEW

**Review by C0rby — COMMENTED** [review]
Unit tests for `NewClaimsSelector` and `NewRegexSelector` would be awesome.

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-owncloud-storage-6</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-9</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-9</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-9</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-9</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-owncloud-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-9</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-9</strong> failed. The build is cancelled...</summary>

**Comment by butonic** [comment]
There actually is a bug: https://github.com/cs3org/reva/pull/1884 No idea why this PR uncovers it ...

**Comment by phil-davis** [comment]
> There actually is a bug: [cs3org/reva#1884](https://github.com/cs3org/reva/pull/1884) No idea why this PR uncovers it ...

This PR changes proxy code - without looking in detail, I guess that the new code struggles to process the body that comes with the `416` "Range Not Satisfiable"  response. But the previous proxy code didn't care?

**Comment by butonic** [comment]
correct. but looking at the code the only thing that could affect the flow is that I am no longer using a fresh context (because I need the context to transport the user from the account_resolver middleware to the actual PolicySelector). But I only wrote unit tests for the new regex and claims selectors ... The acceptance tests still only use the static selector which ALWAYS takes the same route...
🤔

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-ocis-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by ownclouders** [comment]
<details><summary>:boom: Acceptance tests <strong>Core-API-Tests-owncloud-storage-10</strong> failed. The build is cancelled...</summary>

**Comment by butonic** [comment]
- [ ] we should enable login by `ownclouduuid` in https://github.com/owncloud/ocis/blob/1af75187c1d463836623e31855529b2cf6931fe2/proxy/pkg/middleware/account_resolver.go#L65

**Comment by butonic** [comment]
- [ ] we need a cookie to switch users between routes because some requests, especially `config.json` are sent without a Bearer header. To route them properly we have to fall back to a cookie.

**Comment by butonic** [comment]
[tusd] 2021/07/19 14:03:15 event="ChunkWriteComplete" id="eadee146-5057-4433-861e-e7eebdf387db" bytesWritten="7" 
[tusd] 2021/07/19 14:03:15 event="ResponseOutgoing" status="204" method="PATCH" path="/eadee146-5057-4433-861e-e7eebdf387db" requestId="b518997e-976c-4db1-8512-4fc0294c4a51" 
[tusd] 2021/07/19 14:03:15 event="RequestIncoming" method="PATCH" path="/eadee146-5057-4433-861e-e7eebdf387db" requestId="40174af3-b638-4a58-9b77-fab7efb86d4a" 
[tusd] 2021/07/19 14:03:15 event="ResponseOutgoing" status="404" method="PATCH" path="/eadee146-5057-4433-861e-e7eebdf387db" error="upload not found" requestId="40174af3-b638-4a58-9b77-fab7efb86d4a" 
2021/07/19 14:03:15 httputil: ReverseProxy read error during body copy: unexpected EOF
{"level":"error","service":"ocs","error":"{\"code\":404,\"detail\":\"The requested user could not be found\",\"status\":\"Not Found\"}","userid":"Alice","time":"2021-07-19T14:03:16Z","message":"could not get account for user"}
{"level":"error","service":"proxy","error":"{\"id\":\"com.owncloud.api.accounts\",\"code\":401,\"detail\":\"account not found or invalid credentials\",\"status\":\"Unauthorized\"}","query":"login eq 'Alice' and password eq '123456'","time":"2021-07-19T14:03:16Z","message":"error fetching from accounts-service"}
[tusd] 2021/07/19 14:03:16 event="RequestIncoming" method="PATCH" path="/cd44bc26-bb11-48db-b1fa-4749d36a4a60" requestId="b4f8e24a-37b1-44fd-bfd6-178911370024" 
[tusd] 2021/07/19 14:03:16 event="ChunkWriteStart" id="cd44bc26-bb11-48db-b1fa-4749d36a4a60" maxSize="10" offset="0"

**Comment by butonic** [comment]
I found https://github.com/cs3org/reva/pull/1904

Does this PR change the request flow and cause the datagateway to spew errors?!? how?

File: proxy/pkg/middleware/account_resolver.go [hunk 1]
```diff
@@ -89,15 +94,19 @@ func (m accountResolver) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 			return
 		}
 
-		m.logger.Debug().Interface("claims", claims).Interface("user", u).Msgf("associated claims with uuid")
+		// add user to context for selectors
+		ctx = revauser.ContextSetUser(ctx, u)
+		req = req.WithContext(ctx)
```
Comment (wkloucek) [inline proxy/pkg/middleware/account_resolver.go:99]: With this my login fails, because the capabilities endpoint is refused

Comment (wkloucek) [reply proxy/pkg/middleware/account_resolver.go:99]: nevermind, was a fault in my setup

**Comment by wkloucek** [comment]
@butonic I added the cookie logic to this PR

**Comment by butonic** [comment]
Custom claims mapping in https://github.com/owncloud/ocis/pull/2315

**Review by C0rby — APPROVED** [review]

**Pull request merged**